build: upgrade to Go 1.14

otan opened the pull request:
Minor fixes required:
* Upgrade go.mod to prevent build.Import errors.
* Updated instructions to upgrade Go.
* Updated .pb.go files - these are tied with the gofmt of the
  environment and we cannot fix this.
* Update storage use cast DBSlice to DBString or it fails S1016 of
  staticcheck.
* Make lint fail if the import directory is not found.
* Fix acceptance test to include GOROOT or else build.Import fails from
  within the acceptance tests. The acceptance test is weirdly set up to
  run a Go binary from the outside container.

Checklist:
* [x] Adjust version in Docker image ([source](./builder/Dockerfile#L199-L200)).
* [x] Rebuild the Docker image and bump the `version` in `builder.sh` accordingly ([source](./builder.sh#L6)).
* [x] Bump the version in `go-version-check.sh` ([source](./go-version-check.sh)), unless bumping to a new patch release.
* [x] Bump the default installed version of Go in `bootstrap-debian.sh` ([source](./bootstrap/bootstrap-debian.sh#L40-42)).
* [ ] Update the `builder.dockerImage` parameter in the TeamCity [`Cockroach`](https://teamcity.cockroachdb.com/admin/editProject.html?projectId=Cockroach&tab=projectParams) and [`Internal`](https://teamcity.cockroachdb.com/admin/editProject.html?projectId=Internal&tab=projectParams) projects.

(I will do the last step after this has merged)

Resolves #48737.

Release note (general change): Bump the Go version to 1.14.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/50671)
<!-- Reviewable:end -->

petermattis commented:
@bdarnell Wasn't there some problem you discovered with roachtest and go1.14? I just looked and couldn't find an issue. Am I hallucinating?

otan commented:
fails the `worker-fd-fatal` unit test, at least :P

otan commented:
ok, couple of failures.

* `worker-fd-fatal` fails, looks like no error appears. @tbg does this repro for you?
```
------- Stdout: -------
=== RUN   TestClusterMonitor/worker-fd-fatal
02:10:45 cluster.go:348: > /bin/bash -c echo foo && sleep 3& wait
foo
    TestClusterMonitor/worker-fd-fatal: cluster_test.go:251: error details: <nil>
    TestClusterMonitor/worker-fd-fatal: cluster_test.go:252: <nil>
    --- FAIL: TestClusterMonitor/worker-fd-fatal (3.02s)
```
* acceptance tests are failing, and there is not enough log information for me to debug (where are all the artifacts gone?!)

otan commented:
ok figured it out! acceptance tests are broken because of this:

```
agent@gce-agent-20200513-preempt-18:~/work/.go/src/github.com/cockroachdb/cockroach/pkg/acceptance$  env TZ=America/New_York stdbuf -eL -oL ./acceptance.test -l /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/acceptance -test.v -test.timeout 30m
panic: go/build: go list github.com/cockroachdb/cockroach: exit status 2
go: cannot find GOROOT directory: /usr/local/go



goroutine 1 [running]:
github.com/cockroachdb/cockroach/pkg/acceptance/cluster.glob..func1(0xc0001171d0, 0xc0001171d8)
	/go/src/github.com/cockroachdb/cockroach/pkg/acceptance/cluster/dockercluster.go:73 +0xf6
github.com/cockroachdb/cockroach/pkg/acceptance/cluster.init()
	/go/src/github.com/cockroachdb/cockroach/pkg/acceptance/cluster/dockercluster.go:78 +0x163
```

GOPATH / GOROOT is not set on the TeamCity instance that is running the acceptance test. I do wonder if that is because the docker image of these containers is using an older version of Go. It is very strange that they are run by opening a docker instance in the hosted machine.

otan commented:
also a test race failure: https://teamcity.cockroachdb.com/repository/download/Cockroach_UnitTests_Testrace/2040894:id/raw.2.json.txt.tgz!/artifacts/full_output.txt

bdarnell commented:
I diagnosed the `worker-fd-fatal` issue in https://github.com/cockroachdb/cockroach/issues/48737#issuecomment-642234958

tbg commented:
Ah, bummer. That actually gives us a pretty big regression in roachtest then, as there is barely a test that will not call `t.Fatal` on a goroutine. If we merge this as is, we have to expect a much higher roachtest debugging toil rate. There isn't a quick fix, as the roachtest APIs are not conducive to good behavior here and there isn't even a good abstraction to speak of.

tbg commented:
Perhaps the best I can come up with is that we make `*test` use `panic("X")` instead of `Goexit()`.

otan commented:
can we make it do something like

`panic(NewTypedError("fatal"))`, and catch it in recover?

bdarnell commented:
Yeah, panicking with a designated value seems better than using `goexit` anyway. `t.Error(); return` is also equivalent to `t.Fatal` (as long as you return all the way up the stack).

petermattis commented:
> Yeah, panicking with a designated value seems better than using goexit anyway. t.Error(); return is also equivalent to t.Fatal (as long as you return all the way up the stack).

I have this almost working. I'll send a PR shortly.

otan commented:
This is ready!

otan committed:
build: upgrade to Go 1.14

Minor fixes required:
* Upgrade go.mod to prevent build.Import errors.
* Updated instructions to upgrade Go.
* Updated .pb.go files - these are tied with the gofmt of the
  environment and we cannot fix this.
* Update storage use cast DBSlice to DBString or it fails S1016 of
  staticcheck.
* Make lint fail if the import directory is not found.
* Fix acceptance test to include GOROOT or else build.Import fails from
  within the acceptance tests. The acceptance test is weirdly set up to
  run a Go binary from the outside container.

Checklist:
* [x] Adjust version in Docker image ([source](./builder/Dockerfile#L199-L200)).
* [x] Rebuild the Docker image and bump the `version` in `builder.sh` accordingly ([source](./builder.sh#L6)).
* [x] Bump the version in `go-version-check.sh` ([source](./go-version-check.sh)), unless bumping to a new patch release.
* [x] Bump the default installed version of Go in `bootstrap-debian.sh` ([source](./bootstrap/bootstrap-debian.sh#L40-42)).
* [ ] Update the `builder.dockerImage` parameter in the TeamCity [`Cockroach`](https://teamcity.cockroachdb.com/admin/editProject.html?projectId=Cockroach&tab=projectParams) and [`Internal`](https://teamcity.cockroachdb.com/admin/editProject.html?projectId=Internal&tab=projectParams) projects.

(I will do the last step after this has merged)

Release note (general change): Bump the Go version to 1.14.

petermattis reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> module the cast from `DBString` to `DBSlice`. Curious to know more about that as I previously made a change to use the `DBSlice` literal syntax.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50671#-:-MAnFxh9-xo92i8DqILB:b8qy54s)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jlinder, @otan, and @petermattis)

___
*[pkg/storage/rocksdb.go, line 2594 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50671#-MAnFnsX-brxVWeb39it:-MAnFnsX-brxVWeb39iu:b-tm4r16) ([raw file](https://github.com/cockroachdb/cockroach/blob/fffa545e60b462f54a493ee2866aa9e3c9506ec0/pkg/storage/rocksdb.go#L2594)):*
> ```Go
> 		// cannot live past the lifetime of this method, but we're only
> 		// using it to unmarshal the roachpb.
> 		data := cSliceToUnsafeGoBytes(C.DBSlice(result))
> ```

Huh, I'm mildly surprised this works. In https://github.com/cockroachdb/cockroach/pull/45967 we were moving in the opposite direction. I'm curious to know what error `staticcheck` was giving.


<!-- Sent from Reviewable.io -->

petermattis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50671)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jlinder, @otan, and @petermattis)

___
*[pkg/storage/rocksdb.go, line 2594 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50671#-MAnFnsX-brxVWeb39it:-MAnGK5k2Xqv6ywfRDOm:b-3c6ln4) ([raw file](https://github.com/cockroachdb/cockroach/blob/fffa545e60b462f54a493ee2866aa9e3c9506ec0/pkg/storage/rocksdb.go#L2594)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Huh, I'm mildly surprised this works. In https://github.com/cockroachdb/cockroach/pull/45967 we were moving in the opposite direction. I'm curious to know what error `staticcheck` was giving.
</blockquote></details>

I looked up the `staticcheck` error. That's a bit silly. Sometimes I feel these checks go overboard. Oh well.


<!-- Sent from Reviewable.io -->

otan commented:
cc @jordanlewis seeing execgen related build problems in the latest build  (https://teamcity.cockroachdb.com/viewLog.html?buildId=2043512&buildTypeId=Cockroach_UnitTests) and it seems flaky (doesn't fail on same sha on bleeding edge). Is it possible CI has bad leftover state when switching between bad branches containing Eg go files?

jordanlewis commented:
It must be something like that. I don't understand it, though. This always happens during `make generate`, and not any of the other build steps. But `make generate` explicitly regenerates all of the files.

I think that it must be an issue with two steps racing with each other. Is there a way to have the Makefile build certain steps on their own, first, without concurrency?

jordanlewis commented:
I think I see the problem, I'll have a patch hopefully soon. Sorry for the trouble! 👁️ 👄 👁️

jordanlewis commented:
I think the problem is that something is triggering this line before `make execgen` is finished:

```
GOFLAGS= go build -o cockroach -v  -mod=vendor -tags ' gss make x86_64_pc_linux_gnu' -ldflags '-X github.com/cockroachdb/cockroach/pkg/build.typ=development -extldflags "" -X "github.com/cockroachdb/cockroach/pkg/build.tag=v20.2.0-alpha.1-683-gfffa545e60" -X "github.com/cockroachdb/cockroach/pkg/build.rev=fffa545e60b462f54a493ee2866aa9e3c9506ec0" -X "github.com/cockroachdb/cockroach/pkg/build.cgoTargetTriple=x86_64-pc-linux-gnu"   -X "github.com/cockroachdb/cockroach/pkg/build.tag=v20.2.0-alpha.1-683-gfffa545e60" -X "github.com/cockroachdb/cockroach/pkg/build.rev=fffa545e60b462f54a493ee2866aa9e3c9506ec0" -X "github.com/cockroachdb/cockroach/pkg/build.cgoTargetTriple=x86_64-pc-linux-gnu"   -X "github.com/cockroachdb/cockroach/pkg/build.utcTime=2020/06/27 00:33:29"' ./pkg/cmd/cockroach
```

This appears to be the $(COCKROACH) target itself, though. The script here runs `make generate`. What part of the `generate` target triggers the $(COCKROACH) target and why? That's the thing that would need to be gated behind finishing the complete run of `make execgen`.

jordanlewis commented:
Ah, building `settings.html` apparently requires building all of cockroach.

jordanlewis commented:
#50726 will hopefully fix this

otan commented:
bors r=petermattis

it's time!
